core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -1131,11 +1131,28 @@ private SqlNode createSqlWindowBound(RexWindowBound rexWindowBound) {
     private List<SqlNode> toSql(@Nullable RexProgram program, List<RexNode> operandList) {
       final List<SqlNode> list = new ArrayList<>();
       for (RexNode rex : operandList) {
-        list.add(toSql(program, rex));
+        SqlNode node = toSql(program, rex);
+        boolean flag = (rex.getType().getSqlTypeName() == SqlTypeName.BOOLEAN
+            || rex.getType().getSqlTypeName() == SqlTypeName.TINYINT)
+            && rex.getClass().toString().equals(RexLiteral.class.toString())
+            && (((RexLiteral) rex).getTypeName() == SqlTypeName.NULL
+            || ((RexLiteral) rex).getTypeName() == SqlTypeName.DECIMAL);
+        if (flag) {
+          node = castNullType(node, rex.getType());
+        }
+        list.add(node);
```

Comment:
This code should not be here. Your change is specific to CASE 
So kindly move this code to toSql(@Nullable RexProgram program, RexNode rex) method under CASE.

Reply:
For other cases also please add the code in correct place (if required).